components/application-role-mgt/org.wso2.carbon.identity.application.role.mgt/src/main/java/org/wso2/carbon/identity/application/role/mgt/ApplicationRoleManagerImpl.java

```diff
@@ -0,0 +1,244 @@
+import static org.wso2.carbon.identity.application.role.mgt.constants.ApplicationRoleMgtConstants.ErrorMessages.ERROR_CODE_ROLE_NOT_FOUND;
+import static org.wso2.carbon.identity.application.role.mgt.constants.ApplicationRoleMgtConstants.LOCAL_IDP;
+import static org.wso2.carbon.identity.application.role.mgt.util.ApplicationRoleMgtUtils.handleClientException;
+
+/**
+ * Application role management service implementation.
+ */
+public class ApplicationRoleManagerImpl implements ApplicationRoleManager {
+
+    private static final ApplicationRoleManager instance = new ApplicationRoleManagerImpl();
+
+    private ApplicationRoleManagerImpl() {
+
+    }
+
+    public static ApplicationRoleManager getInstance() {
+
+        return instance;
+    }
+
+    private final ApplicationRoleMgtDAO applicationRoleMgtDAO =
+            new CacheBackedApplicationRoleMgtDAOImpl(new ApplicationRoleMgtDAOImpl());
+
+    @Override
+    public ApplicationRole addApplicationRole(ApplicationRole applicationRole)
+            throws ApplicationRoleManagementException {
+
+        String tenantDomain = ApplicationRoleMgtUtils.getTenantDomain();
+        if (StringUtils.isBlank(applicationRole.getRoleName())) {
+            throw handleClientException(ERROR_CODE_INVALID_ROLE_NAME);
+        }
+        boolean existingRole =
+                applicationRoleMgtDAO.isExistingRole(applicationRole.getApplicationId(), applicationRole.getRoleName(),
+                        tenantDomain);
+        if (existingRole) {
+            throw handleClientException(ERROR_CODE_DUPLICATE_ROLE, applicationRole.getRoleName(),
+                    applicationRole.getApplicationId());
+        }
+        // Validate scopes are authorized to given application.
+        ApplicationRoleMgtUtils.validateAuthorizedScopes(applicationRole.getApplicationId(),
+                Arrays.asList(applicationRole.getPermissions()));
+        return applicationRoleMgtDAO.addApplicationRole(applicationRole, tenantDomain);
+    }
+
+    @Override
+    public ApplicationRole updateApplicationRole(String applicationId, String roleId, String newName,
+                                                 List<String> addedScopes, List<String> removedScopes)
+            throws ApplicationRoleManagementException {
+
+        String tenantDomain = ApplicationRoleMgtUtils.getTenantDomain();
+        validateAppRoleId(roleId);
+        if (newName != null) {
+            // Check whether new role name is valid.
+            if (StringUtils.isBlank(newName)) {
+                throw handleClientException(ERROR_CODE_INVALID_ROLE_NAME);
+            }
+            // Check whether new role name is already exists.
+            boolean existingRole =
+                    applicationRoleMgtDAO.isExistingRole(applicationId, newName, tenantDomain);
+            if (existingRole) {
+                throw handleClientException(ERROR_CODE_DUPLICATE_ROLE, newName,
+                        applicationId);
+            }
```

Comment:
Can we refactor this into a private method since this is being used in several places? Basically to have a method to validate the roleName